packages/yew-router/src/router.rs

```diff
@@ -1,232 +1,145 @@
 //! Router Component.
 
-use crate::{
-    agent::{RouteAgentBridge, RouteRequest},
-    route::Route,
-    RouteState, Switch,
-};
-use std::{
-    fmt::{self, Debug, Error as FmtError, Formatter},
-    rc::Rc,
-};
-use yew::{html, virtual_dom::VNode, Component, ComponentLink, Html, Properties, ShouldRender};
-
-/// Any state that can be managed by the `Router` must meet the criteria of this trait.
-pub trait RouterState: RouteState + PartialEq {}
-impl<STATE> RouterState for STATE where STATE: RouteState + PartialEq {}
-
-/// Rendering control flow component.
-///
-/// # Example
-/// ```
-/// use yew::{prelude::*, virtual_dom::VNode};
-/// use yew_router::{router::Router, Switch};
-///
-/// pub enum Msg {}
-///
-/// pub struct Model {}
-/// impl Component for Model {
-///     //...
-/// #   type Message = Msg;
-/// #   type Properties = ();
-/// #   fn create(_: Self::Properties, _link: ComponentLink<Self>) -> Self {
-/// #       Model {}
-/// #   }
-/// #   fn update(&mut self, msg: Self::Message) -> ShouldRender {
-/// #        false
-/// #   }
-/// #   fn change(&mut self, props: Self::Properties) -> ShouldRender {
-/// #        false
-/// #   }
-///
-///     fn view(&self) -> VNode {
-///         html! {
-///         <Router<S>
-///            render = Router::render(|switch: S| {
-///                match switch {
-///                    S::Variant => html!{"variant route was matched"},
-///                }
-///            })
-///         />
-///         }
-///     }
-/// }
-///
-/// #[derive(Switch, Clone)]
-/// enum S {
-///     #[at = "/v"]
-///     Variant,
-/// }
-/// ```
-// TODO, can M just be removed due to not having to explicitly deal with callbacks anymore? - Just get rid of M
-#[derive(Debug)]
-pub struct Router<SW: Switch + Clone + 'static, STATE: RouterState = ()> {
-    switch: Option<SW>,
-    props: Props<STATE, SW>,
-    router_agent: RouteAgentBridge<STATE>,
-}
-
-impl<SW, STATE> Router<SW, STATE>
-where
-    STATE: RouterState,
-    SW: Switch + Clone + 'static,
-{
-    /// Wrap a render closure so that it can be used by the Router.
-    /// # Example
-    /// ```
-    /// # use yew_router::Switch;
-    /// # use yew_router::router::{Router, Render};
-    /// # use yew::{html, Html};
-    /// # #[derive(Switch, Clone)]
-    /// # enum S {
-    /// #     #[at = "/route"]
-    /// #     Variant
-    /// # }
-    /// # pub enum Msg {}
-    ///
-    /// # fn dont_execute() {
-    /// let render: Render<S> = Router::render(|switch: S| -> Html {
-    ///     match switch {
-    ///         S::Variant => html! {"Variant"},
-    ///     }
-    /// });
-    /// # }
-    /// ```
-    pub fn render<F: RenderFn<Router<SW, STATE>, SW> + 'static>(f: F) -> Render<SW, STATE> {
-        Render::new(f)
+use crate::utils::{base_url, build_path_with_base};
+use crate::Routable;
+use gloo::events::EventListener;
+use std::collections::HashMap;
+use std::rc::Rc;
+use yew::prelude::*;
+
+pub struct RenderFn<R>(Rc<dyn Fn(R) -> Html>);
```

Comment:
That generates
```rust
impl <R: Clone> for RenderFn<R>
```
We need to make sure `R` doesn't have the `Clone` bound because we clone the `Rc`, not `R`

Reply:
Ah, is that still an issue? Guess it's fine then